### PR TITLE
registration add

### PR DESCRIPTION
**Pull request by ViktoriyaSynytsya**
Finish registration

### CODE REVIEW

File: page_objects/login_page.js [hunk 1]
```diff
@@ -1,31 +1,62 @@
...
     }
-    async getErrorMessage () {
-        return await this.getErrorElement().getText(); 
-    }   
-    async clearCredentials () {
-         await this.getEmailInput().clear();
-         await this.getPassInput().clear();
+
+
+    getRandomEmail() {
```
Comment (nbabii) [inline page_objects/login_page.js]: the method is not related to the page object functionality, please create class Utils and move it there

File: page_objects/login_page.js [hunk 2]
```diff
@@ -1,31 +1,62 @@
...
+        return this.getRandomString(12);
+    }
+
+    getRandomString(characterLength) {
```
Comment (nbabii) [inline page_objects/login_page.js]: 1. use randomstring library, as we did, https://www.npmjs.com/package/randomstring
2. the method is not related to the page object functionality, please create class Utils and move it there
3. if method name `getRandomString` it should return string, not email

File: specs/registration_spec.js [hunk 3]
```diff
@@ -0,0 +1,19 @@
...
+
+describe('Protractor Demo App Login', function () {
+    it('unsuccessful login', async function () {
+        await indexPage.open();
+        await indexPage.waitForPageLoaded();
+        await indexPage.navigateToLoginPage(); 
+        await loginPage.waitForPageLoaded();
+        expect(await loginPage.getTitle()).toEqual('Login - My Store');
+        await loginPage.sendEmail(loginPage.getRandomEmail());
+        await browser.sleep(5000);
```
Comment (nbabii) [inline specs/registration_spec.js]: remove all sleeps

File: test_random.js [hunk 4]
```diff
@@ -0,0 +1,13 @@
+//const random = require("../node_modules/random-js").Random;
```
Comment (nbabii) [inline test_random.js]: Please use this file for new class Utils

File: specs/registration_spec.js [hunk 5]
```diff
@@ -0,0 +1,19 @@
+let indexPage = require ("../page_objects/index_page")
```
Comment (nbabii) [inline specs/registration_spec.js:1]: File name is `registration_spec.js` but inside we have `unsuccessful login` test. Looks like registration test is missed

**Review by nbabii — CHANGES_REQUESTED** [review]

File: elements/dropdown.element.js [hunk 6]
```diff
@@ -0,0 +1,18 @@
+let BaseElement  = require ("../base/base.element");
+let Button = require("../elements/button.element")
+
+class Dropdown extends BaseElement {
+
+    async open() {
+        await this.protractorElement.click();
+    }
+    async selectItem(optionText) {
+        new Button(element(by.xpath("//option[contains(text(), '"+optionText+"')]")), 'Option').click();
```
Comment (nbabii) [inline elements/dropdown.element.js]: new Button(element(by.xpath(\`//option[contains(text(), "${optionText}")]\`)), 'Option').click();

File: page_objects/login_page.js [hunk 7]
```diff
@@ -1,31 +1,110 @@
 // const { element } = require("protractor");
+const random = require("../node_modules/random-js").Random;
```
Comment (nbabii) [inline page_objects/login_page.js]: user randomstring

File: package.json [hunk 8]
```diff
@@ -10,6 +10,7 @@
         "jasmine": "^3.5.0",
         "jasmine-allure-reporter": "^1.0.2",
         "protractor": "^7.0.0",
+        "random-js": "^2.1.0",
```
Comment (nbabii) [inline package.json:13]: user randomstring

File: page_objects/login_page.js [hunk 9]
```diff
@@ -1,31 +1,110 @@
...
+        await allure.createStep(`Register in ${password}`, async () => {
+            await this.getFirstNameInputForReg().sendKeys(firstName);
+            await this.getLastNameInputForReg().sendKeys(lastName);
+            await this.getPassInput().sendKeys(password);
+            //await this.getFirstNameAddInputForReg().sendKeys(firstNameAdd);
+            //await this.getLastNameAddInputForReg().sendKeys(lastNameAdd);
+            await this.getAddressInputForReg().sendKeys(address);
+            await this.getCityInputForReg().sendKeys(city);
+            await this.getStateDropdownForReg().open();
+            await this.getStateDropdownForReg().selectItem('Hawaii');
```
Comment (nbabii) [inline page_objects/login_page.js]: use variable for state value

File: page_objects/login_page.js [hunk 10]
```diff
@@ -1,31 +1,110 @@
...
+            await browser.sleep(3000);
+            await this.getPostCodeInputForReg().sendKeys(zipcode);
+            await this.getCountryDropdownForReg().open();
+            await this.getCountryDropdownForReg().selectItem('United States');
```
Comment (nbabii) [inline page_objects/login_page.js]: use variable for country

File: test_random.js (hunk 4 above)
Comment (nbabii) [reply test_random.js:1]: use `static` methods

File: page_objects/login_page.js (hunk 7 above)
Comment (nbabii) [reply page_objects/login_page.js]: move under utils class

File: specs/registration_spec.js [hunk 11]
```diff
@@ -0,0 +1,27 @@
...
+        await indexPage.waitForPageLoaded();
+        await indexPage.navigateToLoginPage(); 
+        await loginPage.waitForPageLoaded();
+        expect(await loginPage.getTitle()).toEqual('Login - My Store');
+        await loginPage.sendEmail(loginPage.getRandomEmail());
+        await browser.sleep(2000);
+        await loginPage.createAnAccout();
+        expect(await loginPage.getTextForRegForm()).toEqual('Title');
+        await browser.sleep(2000);
+        await loginPage.getTextFromRegisterForm()
```
Comment (nbabii) [inline specs/registration_spec.js]: extra click

**Review by nbabii — CHANGES_REQUESTED** [review]

File: elements/dropdown.element.js (hunk 6 above)
Comment (ViktoriyaSynytsya) [reply elements/dropdown.element.js]: Fixed

File: specs/registration_spec.js (hunk 11 above)
Comment (ViktoriyaSynytsya) [reply specs/registration_spec.js]: fixed

File: page_objects/login_page.js (hunk 1 above)
Comment (ViktoriyaSynytsya) [reply page_objects/login_page.js]: fixed

File: page_objects/login_page.js (hunk 9 above)
Comment (ViktoriyaSynytsya) [reply page_objects/login_page.js]: fixed

File: page_objects/login_page.js (hunk 10 above)
Comment (ViktoriyaSynytsya) [reply page_objects/login_page.js]: fixed

File: package.json (hunk 8 above)
Comment (ViktoriyaSynytsya) [reply package.json:13]: fixed

File: page_objects/login_page.js (hunk 7 above)
Comment (ViktoriyaSynytsya) [reply page_objects/login_page.js]: fixed

File: test_random.js (hunk 4 above)
Comment (ViktoriyaSynytsya) [reply test_random.js]: fixed

File: specs/registration_spec.js (hunk 3 above)
Comment (ViktoriyaSynytsya) [reply specs/registration_spec.js]: fixed

File: page_objects/login_page.js (hunk 2 above)
Comment (ViktoriyaSynytsya) [reply page_objects/login_page.js]: fixed

File: page_objects/Utils.js [hunk 12]
```diff
@@ -0,0 +1,20 @@
...
+
+class Utils extends BasePage {
+
+    getRandomEmail() {
+        return this.getRandomString(12)+'@gmail.com';;
+    }
+    getRandomString(characterLength) {
+        return randomstring.generate(characterLength);
+    }
+    async clearCredentials() {
```
Comment (nbabii) [inline page_objects/Utils.js]: 1. Utils should contain only methods not related to UI action, some helper tools like data generation, `clearCredentials` should not be here
2. Do not extend BasePage, it is not page object class

File: page_objects/login_page.js [hunk 13]
```diff
@@ -1,31 +1,84 @@
...
+    async createAnAccout() {
+        await this.getSubmButForRegInput().click();
+    }
+    async getTextFromRegisterForm() {
+        await this.getBaseElementForReg().click();
+    }
+    async getTextForRegForm() {
+        return await this.getTitleforExpect().getText();
+    }
+    async getTextForSuccessReg() {
```
Comment (nbabii) [inline page_objects/login_page.js:40]: After the user registered, it is no more login screen showed. This method should be moved to the new page object class. 

If you see the page on UI is changed (the same should be in your source code) page object class should be also changed

File: page_objects/login_page.js [hunk 14]
```diff
@@ -1,31 +1,84 @@
...
+        return await this.getTextforExpect().getText();
     }
+    
+    async registrationForm(firstName, lastName, password, firstNameAdd,
```
Comment (nbabii) [inline page_objects/login_page.js]: The registration form is also part of another page, it is no more login page

File: test_random.js [hunk 15]
```diff
@@ -0,0 +1,30 @@
...
+
```
Comment (nbabii) [inline test_random.js]: Please remove this file

File: unsuccessful_login_spec.js [hunk 16]
```diff
@@ -14,7 +15,7 @@ describe('Protractor Demo App Login', function () {
         await loginPage.login("viktoriasynytsa@gmail.com", "");
         expect(await loginPage.getErrorMessage()).toEqual('Password is required.');
         //await browser.sleep(5000);
-        await loginPage.clearCredentials();
+        await utils.clearCredentials();
```
Comment (nbabii) [inline unsuccessful_login_spec.js]: `clearCredentials ` should be part of the login page, not utils

**Review by nbabii — CHANGES_REQUESTED** [review]

File: page_objects/login_page.js [hunk 17]
```diff
@@ -36,10 +89,58 @@ class LoginPage extends BasePage {
...
-        return  new TextView(element(errorpassword), 'Error about unsuccessful login');
+        return new TextView(element(errorpassword), 'Error about unsuccessful login');
+    }
+    getEmailInputForReg() {
+        return new Input(element(emailforRegLocator), 'email');
+    }
+    getSubmButForRegInput() {
+        return new Button(element(createAnAccountLocator), 'Submit');
+    }
+    getBaseElementForReg() {
```
Comment (nbabii) [inline page_objects/login_page.js]: rename

**Review by nbabii — CHANGES_REQUESTED** [review]

File: page_objects/login_page.js (hunk 17 above)
Comment (ViktoriyaSynytsya) [reply page_objects/login_page.js]: fixed

File: unsuccessful_login_spec.js (hunk 16 above)
Comment (ViktoriyaSynytsya) [reply unsuccessful_login_spec.js]: fixed

File: test_random.js (hunk 15 above)
Comment (ViktoriyaSynytsya) [reply test_random.js]: fixed

File: page_objects/login_page.js (hunk 14 above)
Comment (ViktoriyaSynytsya) [reply page_objects/login_page.js]: fixed

File: page_objects/login_page.js (hunk 13 above)
Comment (ViktoriyaSynytsya) [reply page_objects/login_page.js:40]: fixed

File: page_objects/Utils.js (hunk 12 above)
Comment (ViktoriyaSynytsya) [reply page_objects/Utils.js]: fixed

File: page_objects/accAfterReg_page.js [hunk 18]
```diff
@@ -0,0 +1,15 @@
+let BasePage = require("../base/base.page")
+let TextView = require("../elements/text.view")
+
+let successfulReg = by.css('.info-account')
+
+class Registration extends BasePage {
```
Comment (nbabii) [inline page_objects/accAfterReg_page.js]: The class name already used in other page object

**Review by nbabii — CHANGES_REQUESTED** [review]

File: page_objects/accAfterReg_page.js (hunk 18 above)
Comment (ViktoriyaSynytsya) [reply page_objects/accAfterReg_page.js]: fixed

**Pull request merged**